### DISCUSSION TITLE
Dynamic connections to external database servers

### DISCUSSION THREAD

**Discussion by koiosway** [sqlalchemy/sqlalchemy#7234] [Usage Questions] [answered]
Hi everyone!

I am developing a multi-tenant geospatial application (Python REST API with FastAPI, serving a Single-Page Application) using SQLAlchemy as ORM and PostgreSQL/PostGIS as database. Additionally to the core application tables (users, projects, layers, etc.) that are general to all tenants, I've also decided to use a separate PostgreSQL schema for each tenant where they can import their own data independently (essentially geospatial layers as PostGIS tables).

Besides the core application functionality, there are two use cases I want to implement:

1. Allowing the different tenants to dynamically (at runtime) import geospatial data to their personal schema as well as query/update that data, which seems I can implement using SQLAlchemy Core.

2. Being able to dynamically set up connections to external database servers so that each tenant can potentially use in our application data hosted in their own infrastructure (outside ours), which seems to be more challenging. Basically, the idea would be to dynamically add new database URIs and then be able to query specific tables as if they were part of the application's database. What would be the recommended way to do this using SQLAlchemy?

Creating dynamic engines for every request doesn't seem like a good idea. Perhaps there is a way to use a central store like Redis to cache the different connections and be used efficiently by the different application processes.

I am also aware of the possibility of using the application's database as a sort of "cache" for the external databases, where each "imported" table would be queried and copied to a table in the application's database and then be synchronized periodically. But I was wondering if there is a better and more recommended way to achieve this dynamically and live.

Thanks a lot!

**Comment by zzzeek** [comment, marked as answer]
I would be nervous about launching a multi-tenant application where users can input arbitrary IP numbers and then your server would be connecting to them.    there are ways to make not-well-behaving servers where your connections could get stuck or filled with excess data, etc and you'd have to be very careful to guard against scenarios that can impact your service.     I would probably prefer to build it using some kind of firewall where your app pulls from a local database that was separately updated with the user's tables.   that way you can guarantee performance and you can also join these schemas dynamically using [foreign data wrappers](https://www.postgresql.org/docs/9.5/postgres-fdw.html), that way the whole namespace of tables is available at once.

SQLAlchemy's create_engine() is per-database-server so if you were trying to connect out directly to external DBs you would need to maintain one of these for each remote database.  I'd probably use NullPool so that you dont have idle connections left opened to these remote databases when not in use.  querying from multiple databases at once is also a little less convenient as you have to refer to the right engine at the right time.

i dont know what using redis refers towards, that's a data storage layer.  a database connection isn't really "data" it's a live socket.  while these exist as a file-like object in unix systems I dont think it's really feasible that these can be packaged up and shipped between unrelated processes.

so to sum up, 1. I'd avoid having untrusted IP numbers as callouts in the main application 2. I'd look to see what Postgresql foreign data wrappers can do here.

**Reply by zzzeek** [reply]
also look at [pgbouncer](https://www.pgbouncer.org/) which also might have some features that are useful here.

**Reply by koiosway** [reply]
Many thanks @zzzeek for your helpful and detailed advice, and for pointing out the dangers of arbitrary connections.
I didn't know about Postgresql foreign data wrappers, and it seems like a great solution for this!
Will check out pgbouncer too.
